programs/whirlpool/src/util/token_2022.rs

```diff
@@ -0,0 +1,306 @@
+use anchor_lang::prelude::*;
+use anchor_spl::associated_token::{self, AssociatedToken};
+use anchor_spl::token_2022::spl_token_2022::{
+    self, extension::ExtensionType, instruction::AuthorityType,
+};
+use anchor_spl::token_2022::Token2022;
+use anchor_spl::token_interface::{Mint, TokenAccount};
+use solana_program::program::{invoke, invoke_signed};
+use solana_program::system_instruction::create_account;
+
+use crate::constants::{
+    WP_2022_METADATA_MAX_LEN, WP_2022_METADATA_NAME_PREFIX, WP_2022_METADATA_SYMBOL,
+    WP_2022_METADATA_URI_BASE,
+};
+use crate::state::*;
+
+pub fn initialize_position_mint_2022<'info>(
+    position_mint: &Signer<'info>,
+    funder: &Signer<'info>,
+    position: &Account<'info, Position>,
+    system_program: &Program<'info, System>,
+    token_2022_program: &Program<'info, Token2022>,
+    with_token_metadata_extension: bool,
+) -> Result<()> {
+    let space = ExtensionType::try_calculate_account_len::<spl_token_2022::state::Mint>(
+        if with_token_metadata_extension {
+            &[
+                ExtensionType::MintCloseAuthority,
+                ExtensionType::MetadataPointer,
+            ]
+        } else {
+            &[ExtensionType::MintCloseAuthority]
+        },
+    )?;
+
+    let lamports = Rent::get()?.minimum_balance(if with_token_metadata_extension {
+        // we need to add rent for TokenMetadata extension to reallocate space
+        space + WP_2022_METADATA_MAX_LEN
+    } else {
+        space
+    });
+
+    let authority = position;
+
+    // create account
+    invoke(
+        &create_account(
+            funder.key,
+            position_mint.key,
+            lamports,
+            space as u64,
+            token_2022_program.key,
+        ),
+        &[
+            funder.to_account_info(),
+            position_mint.to_account_info(),
+            token_2022_program.to_account_info(),
+            system_program.to_account_info(),
+        ],
+    )?;
+
+    // initialize MintCloseAuthority extension
+    // authority: Position account (PDA)
+    invoke(
+        &spl_token_2022::instruction::initialize_mint_close_authority(
+            token_2022_program.key,
+            position_mint.key,
+            Some(&authority.key()),
+        )?,
+        &[
+            position_mint.to_account_info(),
+            authority.to_account_info(),
+            token_2022_program.to_account_info(),
+        ],
+    )?;
+
+    if with_token_metadata_extension {
+        // initialize MetadataPointer extension
+        // authority: None
```

Comment:
if we are setting the UpdateAuthority on the TokenMetadata anyways, is there any downside to also setting the authority here?

Reply:
I don't think there is any clear downside.
I think `None` is "a bit" safer than `Some(auth)` here.

Do you think setting WHIRLPOOL_NFT_UPDATE_AUTH would avoid our future regret ?

background:
I thought that there is no possibility of modification, and so I set Authority to None.
The case where MetadataPointer refers to something other than Mint itself would be if we stop using the TokenMetadata extension and start using Metaplex TokenMetadata again, which I thought was unlikely to happen.

As for the metadata, we have made changes to add “DO NO BURN” in the past. So we need to be able to change it.